### PR TITLE
[RFC] Licensescanner

### PR DESCRIPTION
**Pull request by rhubert**
I started working on adding a license scanner to bob. After some research it seams that [fossology](https://www.fossology.org/) is the most active developed scanner producing SPDX files which can be post-processed by the tools you can get from https://spdx.org/.

But there is no stable WebAPI to work with fossology. There are some [ideas](https://github.com/fossology/fossology/pull/837) and [issues](https://github.com/fossology/fossology/issues/592) about this so hopefully this will change in the future.

As of today there are some commandline tools available to upload files to fossology. I'm using ssh/scp to upload and execute the commandline tools. Afterwards I use curl to poll for the spdx. All in all it's somehow working...

Installing fossology should be possible on some linux machines, but the use of the prebuild docker image is recommended. The official image has no ssh login. That's why I added some docker-files in the contrib folder to build a custom fossology image with ssh login. I also added a Postgres server for fossology where they store the output of the scanner to have a faster second run on the same file.

Now that I have a spdx file I'm unsure how to proceed with this file? Add it to the audit file? Or put it as separate file to the `meta` folder? 

Scanning for licenses takes a (very) long time. I don't think it's a task which should be done on every local build. It's a task for a (long) running build on a separate machine. But maybe it makes sense to upload the spdx and download it during local builds? (ATM I can't imagine why I should ever need this locally, but maybe there is a use case? 
I'm also thinking about howto exclude some 'know' proprietary software from getting scanned. Maybe using a special `metaEnvironment` variable?

I think it may also be useful to add the licenseScanner functionality to the jenkins builder.This would require to make the scanner configurable (servername, port, login, ... )???

So many questions... 🤔 

I also added some special metaEnv LICENSE_FILE and LICENSE_CHECKSUM functionality and moved to spdx identifiers...

### CODE REVIEW

**Comment by jkloetzke** [comment]
These are just my initial thoughts. I have the feeling this will get a rather long PR. ;-) I think there are two distinct features that can be addressed here. The fist is the actual integration of license scanners and the second is to express assertions about the source code of a recipe.

# License scanners

Having them as plug-ins is generally a good idea. It would be better if the scanner plugin would generate some shell script snippet that can be appended to the checkout script. This way we would implicitly gain Jenkins support too.

The result should be put in the `meta` folder. That was it's intention in the first place. I'm not sure though how transitivity should be handled. Not at all -- sounds bad? Copy license reports of dependencies into result artifact? Or even combine them? The latter would require to strictly specify the output format which feels not right.

Given the fact that there is no WebAPI for Fossology it fells a bit premature to include it as a built-in scanner. At least the scanner should be made configurable. This leads to another open issue that I was thinking about for a while: passing plugins configuration options from `default.yaml`. The fossology plugin would be classic example where the user has to specify if the tools are installed locally or if ssh is needed. In case of ssh a username/password is needed.

Running these scanners *must* be optional because not everybody may have the required tools installed.

# Expressing assertions about source code

I'm not really inclined to give certain `metaEnvironment` variables a meaning in Bob. If they have special semantics they should be keywords on their own right. If appropriate they could still contribute to the meta environment.

The LICENSE_FILE and LICENSE_CHECKSUM variables also miss some flexibility. What if the package is multi-licensed? We must support checking multiple files. IMHO this fells more like an `assert` statement. Just a quick sketch:

```yaml
checkoutAssert:
    file: COPYING
    digestSHA1: 012345...
```

or more complex:

```yaml
checkoutAssert:
    -   file: main.c
        start: 2
        end: 5
        digestSHA1: 012345678...
    -   file: cli.c
        end: 10
        digestSHA1: abcdef...
```

Do you have any idea how this could be combined with the scanners? In the recipe I can express which licenses I'm expecting. This could be compared with the scanner result if the assumption in the recipe still holds.

File: LICENSE [hunk 1]
```diff
@@ -673,3 +673,11 @@ the library.  If this is what you want to do, use the GNU Lesser General
 Public License instead of this License.  But first, please read
 <http://www.gnu.org/philosophy/why-not-lgpl.html>.
 
+-----------------------------------------------------------------------------
+Note:
+Individual files contain the following tag instead of the full license text.
+
+        SPDX-License-Identifier: GPL-3.0
```
[the remark this answers is not in the document]
Comment (rhubert) [reply LICENSE]: 😟 You're right. Corrected to [GPL-3.0-or-later](https://spdx.org/licenses/GPL-3.0-or-later.html) and moved this commit to a separate PR. So we can focus on the real changes in this PR

**Comment by rhubert** [comment]
### License scanners

For configuring the scanner / plugins a plugin could also look for a plugin_config.yaml file?

I think we need to pass the scan-result of dependencies to the upstream recipe but not combine them. There is  a [tool](https://github.com/spdx/tools) available supporting merge of spdx documents, but this needs `java` and I don't like to bring in another tool. The `meta` folder of a upstream would contain license files of all the downstreams, named using recipe.path and recipe.name?
The other option is to define a destination where licenseFiles are storred. Could be a special `archive` type? This could also be used as licenseFile cache if the source hash is also taken into account?

Running the scanner is optional. I added `--licenseScan` argument to build/dev. For jenkins this option is missing.

### Expressing assertions about source code

I really like your idea to handle this with a `checkoutAssert` keyword. This would also avoid the usage of these policy. I'll rework the PR.

To compate the scanner result with the `checkoutAssert` I'd a method to the licenseScanner plugin where the output of the scanner is parsed and the check can be performed. In case of fossology/spdx it's a xml like description. Other scanners may have to implement something different.

**Comment by jkloetzke** [comment]
I would suggest to split this into two PRs: adding license scanners and source code assertions. They are related but can be implemented independent of each other.

> For configuring the scanner / plugins a plugin could also look for a plugin_config.yaml file?

That's not a good idea. Plugins must not read files on their own because otherwise Bob cannot invalidate the parsing cache if these files change. I also like the idea that plugin options could gain the global/user/project hierarchy. I'll prepare a PR because it can be implemented quite easily IMHO.

> I think we need to pass the scan-result of dependencies to the upstream recipe but not combine them.

Agreed.

> The meta folder of a upstream would contain license files of all the downstreams, named using recipe.path and recipe.name?

I would put everything in a dedicated folder in `meta`, e.g. `meta/licenses`. The indivital files can be named like PackageName-BuildId.extension. There might be multiple files with the same package name because of different variants. Not sure what you mean by recipe.path, though...

> This could also be used as licenseFile cache if the source hash is also taken into account?

Caching could be part of the license scanner infrastructure. Bob could maintain that cache and call the plugin(s) only if the source code hash was not scanned by a plugin yet.

**Comment by jkloetzke** [comment]
It's possible now to pass settings from `default.yaml` to a plugin. See http://bob-build-tool.readthedocs.io/en/latest/manual/extending.html#plugin-settings. This could be used for an example Fossology license scanner plugin to configure it as needed.

**Pull request closed without merging**